File: cpp/plan.cc
#include "plan.h"

#include <algorithm>
#include <array>
#include <limits>
#include <string>

namespace aginfer::internal {
namespace {

constexpr std::array<char, 8> kPlanMagic{'A', 'I', 'M', 'P', 'L', 'N', '1', '\0'};

Status Error(const std::string& message) {
  return Status(StatusCode::kCorruptPackage, message);
}

bool AllZero(const std::uint8_t* data, std::size_t size) {
  return std::all_of(data, data + size, [](std::uint8_t value) { return value == 0; });
}

bool RegionValid(std::size_t total, std::uint64_t offset,
                 std::uint64_t count, std::size_t item_size) {
  if (offset > total || count > std::numeric_limits<std::size_t>::max() / item_size) return false;
  const std::size_t bytes = static_cast<std::size_t>(count) * item_size;
  return bytes <= total - static_cast<std::size_t>(offset);
}

bool RangeValid(std::uint32_t total, std::uint32_t first, std::uint32_t count) {
  return first <= total && count <= total - first;
}

std::uint64_t DTypeSize(std::uint32_t dtype) {
  switch (static_cast<PlanDType>(dtype)) {
    case PlanDType::kFp32:
    case PlanDType::kInt32: return 4;
    case PlanDType::kFp16:
    case PlanDType::kBf16: return 2;
    case PlanDType::kFp8E4M3: return 1;
  }
  return 0;
}

Status ValidateTensor(const ParsedPlan& plan, const PlanTensor& tensor,
                      std::uint32_t index) {
  const auto name = plan.String(tensor.name_offset);
  if (name.empty()) return Error("execution plan tensor " + std::to_string(index) + " has an invalid name");
  const std::uint64_t item_size = DTypeSize(tensor.dtype);
  if (item_size == 0) return Error("execution plan tensor " + std::to_string(index) + " has an invalid dtype");
  if (tensor.location != static_cast<std::uint32_t>(PlanLocation::kHost) &&
      tensor.location != static_cast<std::uint32_t>(PlanLocation::kDevice))
    return Error("execution plan tensor " + std::to_string(index) + " has an invalid location");
  if (tensor.io_kind != static_cast<std::uint32_t>(PlanIoKind::kInput) &&
      tensor.io_kind != static_cast<std::uint32_t>(PlanIoKind::kOutput))
    return Error("execution plan tensor " + std::to_string(index) + " has an invalid I/O kind");
  if (tensor.rank == 0 || tensor.rank > kMaxTensorRank || tensor.flags != 0)
    return Error("execution plan tensor " + std::to_string(index) + " has an invalid rank or flags");
  std::uint64_t last_element = 0;
  for (std::uint32_t dimension = 0; dimension < tensor.rank; ++dimension) {
    if (tensor.shape[dimension] <= 0 || tensor.stride[dimension] <= 0)
      return Error("execution plan tensor " + std::to_string(index) + " has a non-positive shape or stride");
    const auto extent = static_cast<std::uint64_t>(tensor.shape[dimension] - 1);
    const auto stride = static_cast<std::uint64_t>(tensor.stride[dimension]);
    if (extent != 0 && stride > (std::numeric_limits<std::uint64_t>::max() - last_element) / extent)
      return Error("execution plan tensor size overflows");
    last_element += extent * stride;
  }
  for (std::uint32_t dimension = tensor.rank; dimension < kMaxTensorRank; ++dimension) {
    if (tensor.shape[dimension] != 0 || tensor.stride[dimension] != 0)
      return Error("execution plan tensor has non-zero dimensions beyond its rank");
  }
  if (last_element == std::numeric_limits<std::uint64_t>::max() ||
      last_element + 1 > std::numeric_limits<std::uint64_t>::max() / item_size ||
      tensor.byte_size < (last_element + 1) * item_size)
    return Error("execution plan tensor byte size is too small");
  return Status::Ok();
}

}  // namespace

std::string_view ParsedPlan::String(std::uint32_t offset) const {
  if (header == nullptr || offset >= header->string_bytes) return {};
  const char* begin = strings + offset;
  const char* end = strings + header->string_bytes;
  const char* terminator = std::find(begin, end, '\0');
  if (terminator == end) return {};
  return std::string_view(begin, static_cast<std::size_t>(terminator - begin));
}

const PlanProfile* ParsedPlan::FindProfile(std::string_view name) const {
  if (header == nullptr) return nullptr;
  for (std::uint32_t index = 0; index < header->profile_count; ++index) {
    if (String(profiles[index].name_offset) == name) return &profiles[index];
  }
  return nullptr;
}

std::uint32_t ParsedPlan::ProfileIndex(const PlanProfile* profile) const {
  return static_cast<std::uint32_t>(profile - profiles);
}

Status ParsePlan(const std::uint8_t* data, std::size_t size,
                 std::uint32_t expected_arch, std::uint64_t weight_bytes,
                 ParsedPlan* output) {
  if (data == nullptr || output == nullptr || size < sizeof(PlanHeader))
    return Error("execution plan is smaller than its fixed header");
  const auto* header = reinterpret_cast<const PlanHeader*>(data);
  if (!std::equal(kPlanMagic.begin(), kPlanMagic.end(), header->magic))
    return Error("bad AIM execution plan magic");
  if (header->schema_major != kPlanSchemaMajor || header->header_size != sizeof(PlanHeader))
    return Error("unsupported AIM execution plan schema");
  if (header->arch != expected_arch)
    return Error("execution plan architecture does not match its AIM variant");
  if (header->file_size != size || !AllZero(header->reserved, sizeof(header->reserved)))
    return Error("invalid execution plan size or reserved bytes");
  if (header->profile_count == 0 || header->tensor_count == 0 || header->launch_count == 0)
    return Error("execution plan must contain profiles, tensors, and launches");
  if (!RegionValid(size, header->profiles_offset, header->profile_count, sizeof(PlanProfile)) ||
      !RegionValid(size, header->tensors_offset, header->tensor_count, sizeof(PlanTensor)) ||
      !RegionValid(size, header->launches_offset, header->launch_count, sizeof(PlanLaunch)) ||
      !RegionValid(size, header->arguments_offset, header->argument_count, sizeof(PlanArgument)) ||
      !RegionValid(size, header->strings_offset, header->string_bytes, 1) || header->string_bytes == 0 ||
      header->strings_offset + header->string_bytes != size)
    return Error("execution plan contains an invalid table region");
  const std::uint64_t expected_tensors = header->profiles_offset +
      static_cast<std::uint64_t>(header->profile_count) * sizeof(PlanProfile);
  const std::uint64_t expected_launches = header->tensors_offset +
      static_cast<std::uint64_t>(header->tensor_count) * sizeof(PlanTensor);
  const std::uint64_t expected_arguments = header->launches_offset +
      static_cast<std::uint64_t>(header->launch_count) * sizeof(PlanLaunch);
  const std::uint64_t expected_strings = header->arguments_offset +
      static_cast<std::uint64_t>(header->argument_count) * sizeof(PlanArgument);
  if (header->profiles_offset != sizeof(PlanHeader) || header->tensors_offset != expected_tensors ||
      header->launches_offset != expected_launches || header->arguments_offset != expected_arguments ||
      header->strings_offset != expected_strings)
    return Error("execution plan tables are not in canonical order");

  ParsedPlan parsed;
  parsed.data = data;
  parsed.size = size;
  parsed.header = header;
  parsed.profiles = reinterpret_cast<const PlanProfile*>(data + header->profiles_offset);
  parsed.tensors = reinterpret_cast<const PlanTensor*>(data + header->tensors_offset);
  parsed.launches = reinterpret_cast<const PlanLaunch*>(data + header->launches_offset);
  parsed.arguments = reinterpret_cast<const PlanArgument*>(data + header->arguments_offset);
  parsed.strings = reinterpret_cast<const char*>(data + header->strings_offset);
  if (parsed.strings[0] != '\0') return Error("execution plan string table has an invalid sentinel");

  std::uint32_t next_tensor = 0;
  std::uint32_t next_launch = 0;
  for (std::uint32_t profile_index = 0; profile_index < header->profile_count; ++profile_index) {
    const PlanProfile& profile = parsed.profiles[profile_index];
    if (parsed.String(profile.name_offset).empty() || profile.first_tensor != next_tensor ||
        profile.first_launch != next_launch || profile.tensor_count == 0 || profile.launch_count == 0 ||
        !RangeValid(header->tensor_count, profile.first_tensor, profile.tensor_count) ||
        !RangeValid(header->launch_count, profile.first_launch, profile.launch_count) ||
        (profile.flags & ~1U) != 0 || !AllZero(profile.reserved, sizeof(profile.reserved)))
      return Error("execution plan contains an invalid profile record");
    for (std::uint32_t previous = 0; previous < profile_index; ++previous) {
      if (parsed.String(parsed.profiles[previous].name_offset) == parsed.String(profile.name_offset))
        return Error("execution plan contains duplicate profile names");
    }
    next_tensor += profile.tensor_count;
    next_launch += profile.launch_count;
  }
  if (next_tensor != header->tensor_count || next_launch != header->launch_count)
    return Error("execution plan profile ranges do not cover their tables");

  for (std::uint32_t index = 0; index < header->tensor_count; ++index) {
    const Status status = ValidateTensor(parsed, parsed.tensors[index], index);
    if (!status.ok()) return status;
  }
  for (std::uint32_t profile_index = 0; profile_index < header->profile_count; ++profile_index) {
    const PlanProfile& profile = parsed.profiles[profile_index];
    std::uint32_t input_count = 0;
    std::uint32_t output_count = 0;
    for (std::uint32_t index = 0; index < profile.tensor_count; ++index) {
      const PlanTensor& tensor = parsed.tensors[profile.first_tensor + index];
      if (tensor.io_kind == static_cast<std::uint32_t>(PlanIoKind::kInput)) ++input_count;
      else ++output_count;
      for (std::uint32_t previous = 0; previous < index; ++previous) {
        const PlanTensor& other = parsed.tensors[profile.first_tensor + previous];
        if (parsed.String(other.name_offset) == parsed.String(tensor.name_offset))
          return Error("execution plan profile contains duplicate tensor names");
      }
    }
    if (input_count == 0 || output_count == 0)
      return Error("execution plan profile must contain inputs and outputs");
  }

  std::uint32_t next_argument = 0;
  for (std::uint32_t index = 0; index < header->launch_count; ++index) {
    const PlanLaunch& launch = parsed.launches[index];
    if (launch.profile_index >= header->profile_count || parsed.String(launch.kernel_name_offset).empty() ||
        launch.first_argument != next_argument ||
        !RangeValid(header->argument_count, launch.first_argument, launch.argument_count) ||
        launch.grid[0] == 0 || launch.grid[1] == 0 || launch.grid[2] == 0 ||
        launch.block[0] == 0 || launch.block[1] == 0 || launch.block[2] == 0 ||
        static_cast<std::uint64_t>(launch.block[0]) * launch.block[1] * launch.block[2] > 1024 ||
        launch.flags != 0 || !AllZero(launch.reserved, sizeof(launch.reserved)))
      return Error("execution plan contains an invalid launch record");
    const PlanProfile& profile = parsed.profiles[launch.profile_index];
    if (index < profile.first_launch || index >= profile.first_launch + profile.launch_count)
      return Error("execution plan launch belongs to the wrong profile");
    for (std::uint32_t arg_index = 0; arg_index < launch.argument_count; ++arg_index) {
      const PlanArgument& argument = parsed.arguments[launch.first_argument + arg_index];
      if (argument.reserved != 0) return Error("execution plan argument has non-zero reserved data");
      switch (static_cast<PlanArgKind>(argument.kind)) {
        case PlanArgKind::kTensor:
          if (argument.index < profile.first_tensor ||
              argument.index >= profile.first_tensor + profile.tensor_count ||
              parsed.tensors[argument.index].location != static_cast<std::uint32_t>(PlanLocation::kDevice))
            return Error("execution plan launch references an invalid device tensor");
          if (argument.offset != 0 || argument.value != 0)
            return Error("execution plan tensor argument has unexpected payload data");
          break;
        case PlanArgKind::kScalarU32:
        case PlanArgKind::kScalarF32:
          if (argument.index != 0 || argument.offset != 0 ||
              argument.value > std::numeric_limits<std::uint32_t>::max())
            return Error("execution plan scalar argument is out of range");
          break;
        case PlanArgKind::kArenaOffset:
          if (argument.index != 0 || argument.value != 0 || argument.offset >= header->arena_bytes)
            return Error("execution plan arena argument is out of bounds");
          break;
        case PlanArgKind::kWeightsOffset:
          if (argument.index != 0 || argument.value != 0 || argument.offset >= weight_bytes)
            return Error("execution plan weight argument is out of bounds");
          break;
        default: return Error("execution plan contains an unknown argument kind");
      }
    }
    next_argument += launch.argument_count;
  }
  if (next_argument != header->argument_count)
    return Error("execution plan launch ranges do not cover the argument table");
  *output = parsed;
  return Status::Ok();
}

}  // namespace aginfer::internal

File: src/aginfer/plan.py
from __future__ import annotations

import math
import struct
from dataclasses import dataclass
from typing import Any

from .errors import FormatError, ValidationError
from .schema import CudaArch

PLAN_MAGIC = b"AIMPLN1\0"
PLAN_SCHEMA_MAJOR = 1
PLAN_SCHEMA_MINOR = 0
MAX_TENSOR_RANK = 8

PLAN_HEADER_STRUCT = struct.Struct("<8sHHIIIIIIIQQQQQQQQ24s")
PROFILE_STRUCT = struct.Struct("<IIIIII40s")
TENSOR_STRUCT = struct.Struct("<IIIIIIQ8q8q")
LAUNCH_STRUCT = struct.Struct("<IIIIIIIIIIII16s")
ARG_STRUCT = struct.Struct("<IIQQQ")

assert PLAN_HEADER_STRUCT.size == 128
assert PROFILE_STRUCT.size == 64
assert TENSOR_STRUCT.size == 160
assert LAUNCH_STRUCT.size == 64
assert ARG_STRUCT.size == 32

DTYPE_IDS = {"F32": 1, "F16": 2, "BF16": 3, "F8_E4M3": 4, "I32": 5}
DTYPE_SIZES = {"F32": 4, "F16": 2, "BF16": 2, "F8_E4M3": 1, "I32": 4}
LOCATION_IDS = {"host": 1, "device": 2}
IO_IDS = {"input": 1, "output": 2}
ARG_IDS = {"tensor": 1, "scalar_u32": 2, "scalar_f32": 3, "arena_offset": 4, "weights_offset": 5}


@dataclass(frozen=True)
class CompiledPlan:
    data: bytes
    profiles: tuple[str, ...]
    input_names: tuple[str, ...]
    output_names: tuple[str, ...]
    launch_count: int


class _StringTable:
    def __init__(self) -> None:
        self.data = bytearray(b"\0")
        self.offsets: dict[str, int] = {"": 0}

    def add(self, value: str, label: str) -> int:
        if not isinstance(value, str) or not value or "\0" in value:
            raise ValidationError(f"{label} must be a non-empty string without NUL bytes")
        encoded = value.encode("utf-8")
        if len(encoded) > 4096:
            raise ValidationError(f"{label} is too long")
        if value not in self.offsets:
            self.offsets[value] = len(self.data)
            self.data.extend(encoded)
            self.data.append(0)
        return self.offsets[value]


def compile_execution_plan(plan: dict[str, Any], arch: CudaArch, *, weight_size: int) -> CompiledPlan:
    if not isinstance(plan, dict):
        raise ValidationError("execution plan must be an object")
    if plan.get("cuda_arch") != arch.name_string:
        raise ValidationError(f"execution plan target must be exactly {arch.name_string}")
    arena_bytes = _unsigned(plan.get("arena_bytes"), "arena_bytes")
    workspace_bytes = _unsigned(plan.get("workspace_bytes"), "workspace_bytes")
    dispatches = plan.get("shape_dispatch")
    if not isinstance(dispatches, list) or not dispatches:
        raise ValidationError("execution plan needs a non-empty shape_dispatch list")
    graph_profiles = _validate_graph_templates(plan.get("cuda_graph_templates"), dispatches)

    strings = _StringTable()
    profile_records: list[tuple[int, int, int, int, int, int]] = []
    tensor_records: list[tuple[int, int, int, int, int, int, int, tuple[int, ...], tuple[int, ...]]] = []
    launch_records: list[tuple[int, int, int, int, tuple[int, int, int], tuple[int, int, int], int, int]] = []
    argument_records: list[tuple[int, int, int, int]] = []
    profile_names: list[str] = []
    first_input_names: tuple[str, ...] = ()
    first_output_names: tuple[str, ...] = ()

    for profile_index, dispatch in enumerate(dispatches):
        if not isinstance(dispatch, dict):
            raise ValidationError(f"shape_dispatch[{profile_index}] must be an object")
        profile = dispatch.get("profile")
        profile_offset = strings.add(profile, f"shape_dispatch[{profile_index}].profile")
        if profile in profile_names:
            raise ValidationError(f"duplicate execution profile: {profile}")
        profile_names.append(profile)

        first_tensor = len(tensor_records)
        names: dict[str, int] = {}
        device_names: set[str] = set()
        input_names: list[str] = []
        output_names: list[str] = []
        for io_name, io_kind, collected_names in (
            ("inputs", "input", input_names),
            ("outputs", "output", output_names),
        ):
            specs = dispatch.get(io_name)
            if not isinstance(specs, list) or not specs:
                raise ValidationError(f"profile {profile}: {io_name} must be a non-empty list")
            for item_index, spec in enumerate(specs):
                record = _compile_tensor(
                    spec,
                    io_kind,
                    strings,
                    f"profile {profile} {io_name}[{item_index}]",
                )
                name = spec["name"]
                if name in names:
                    raise ValidationError(f"profile {profile}: duplicate tensor name {name}")
                names[name] = len(tensor_records)
                if record[2] == LOCATION_IDS["device"]:
                    device_names.add(name)
                collected_names.append(name)
                tensor_records.append(record)
        if profile_index == 0:
            first_input_names = tuple(input_names)
            first_output_names = tuple(output_names)

        first_launch = len(launch_records)
        launches = dispatch.get("launches")
        if not isinstance(launches, list) or not launches:
            raise ValidationError(f"profile {profile}: launches must be a non-empty list")
        for launch_index, launch in enumerate(launches):
            if not isinstance(launch, dict):
                raise ValidationError(f"profile {profile} launch {launch_index} must be an object")
            kernel_offset = strings.add(launch.get("kernel"), f"profile {profile} launch {launch_index}.kernel")
            grid = _dimensions(launch.get("grid"), f"profile {profile} launch {launch_index}.grid")
            block = _dimensions(launch.get("block"), f"profile {profile} launch {launch_index}.block")
            if math.prod(block) > 1024:
                raise ValidationError(f"profile {profile} launch {launch_index}: block has more than 1024 threads")
            shared_bytes = _unsigned(launch.get("shared_bytes", 0), "shared_bytes", maximum=2**32 - 1)
            arguments = launch.get("arguments", [])
            if not isinstance(arguments, list):
                raise ValidationError(f"profile {profile} launch {launch_index}.arguments must be a list")
            first_arg = len(argument_records)
            for arg_index, argument in enumerate(arguments):
                argument_records.append(
                    _compile_argument(
                        argument,
                        names,
                        device_names,
                        arena_bytes,
                        weight_size,
                        f"profile {profile} launch {launch_index} argument {arg_index}",
                    )
                )
            launch_records.append(
                (kernel_offset, profile_index, first_arg, len(arguments), grid, block, shared_bytes, 0)
            )
        flags = 1 if profile in graph_profiles else 0
        profile_records.append(
            (profile_offset, first_tensor, len(tensor_records) - first_tensor, first_launch, len(launches), flags)
        )

    profiles_offset = PLAN_HEADER_STRUCT.size
    tensors_offset = profiles_offset + len(profile_records) * PROFILE_STRUCT.size
    launches_offset = tensors_offset + len(tensor_records) * TENSOR_STRUCT.size
    args_offset = launches_offset + len(launch_records) * LAUNCH_STRUCT.size
    strings_offset = args_offset + len(argument_records) * ARG_STRUCT.size
    file_size = strings_offset + len(strings.data)
    header = PLAN_HEADER_STRUCT.pack(
        PLAN_MAGIC,
        PLAN_SCHEMA_MAJOR,
        PLAN_SCHEMA_MINOR,
        PLAN_HEADER_STRUCT.size,
        int(arch),
        len(profile_records),
        len(tensor_records),
        len(launch_records),
        len(argument_records),
        len(strings.data),
        arena_bytes,
        workspace_bytes,
        profiles_offset,
        tensors_offset,
        launches_offset,
        args_offset,
        strings_offset,
        file_size,
        b"\0" * 24,
    )
    output = bytearray(header)
    for record in profile_records:
        output.extend(PROFILE_STRUCT.pack(*record, b"\0" * 40))
    for name, dtype, location, io_kind, rank, flags, byte_size, shape, stride in tensor_records:
        padded_shape = shape + (0,) * (MAX_TENSOR_RANK - rank)
        padded_stride = stride + (0,) * (MAX_TENSOR_RANK - rank)
        output.extend(
            TENSOR_STRUCT.pack(name, dtype, location, io_kind, rank, flags, byte_size, *padded_shape, *padded_stride)
        )
    for kernel, profile, first_arg, arg_count, grid, block, shared, flags in launch_records:
        output.extend(
            LAUNCH_STRUCT.pack(
                kernel,
                profile,
                first_arg,
                arg_count,
                *grid,
                *block,
                shared,
                flags,
                b"\0" * 16,
            )
        )
    for kind, index, offset, value in argument_records:
        output.extend(ARG_STRUCT.pack(kind, index, offset, value, 0))
    output.extend(strings.data)
    if len(output) != file_size:
        raise AssertionError("internal execution plan size mismatch")
    return CompiledPlan(bytes(output), tuple(profile_names), first_input_names, first_output_names, len(launch_records))


def inspect_plan(data: bytes) -> dict[str, int]:
    """Validate the fixed Plan v1 header and return its summary fields."""
    if len(data) < PLAN_HEADER_STRUCT.size:
        raise FormatError("execution plan is smaller than its fixed header")
    values = PLAN_HEADER_STRUCT.unpack_from(data)
    if values[0] != PLAN_MAGIC:
        raise FormatError("bad AIM execution plan magic")
    if values[1] != PLAN_SCHEMA_MAJOR or values[3] != PLAN_HEADER_STRUCT.size:
        raise FormatError("unsupported AIM execution plan schema")
    if values[17] != len(data):
        raise FormatError("execution plan size mismatch")
    if any(values[18]):
        raise FormatError("non-zero execution plan reserved bytes")
    return {
        "arch": values[4],
        "profile_count": values[5],
        "tensor_count": values[6],
        "launch_count": values[7],
        "argument_count": values[8],
        "arena_bytes": values[10],
        "workspace_bytes": values[11],
    }


def _compile_tensor(
    spec: Any,
    io_kind: str,
    strings: _StringTable,
    label: str,
) -> tuple[int, int, int, int, int, int, int, tuple[int, ...], tuple[int, ...]]:
    if not isinstance(spec, dict):
        raise ValidationError(f"{label} must be an object")
    name_offset = strings.add(spec.get("name"), f"{label}.name")
    dtype_name = spec.get("dtype")
    if dtype_name not in DTYPE_IDS:
        raise ValidationError(f"{label}.dtype must be one of {', '.join(DTYPE_IDS)}")
    location_name = spec.get("location", "device")
    if location_name not in LOCATION_IDS:
        raise ValidationError(f"{label}.location must be host or device")
    shape_value = spec.get("shape")
    if not isinstance(shape_value, list) or not 1 <= len(shape_value) <= MAX_TENSOR_RANK:
        raise ValidationError(f"{label}.shape must contain 1 to {MAX_TENSOR_RANK} dimensions")
    shape = tuple(_positive_int(item, f"{label}.shape") for item in shape_value)
    stride_value = spec.get("stride")
    if stride_value is None:
        stride = _contiguous_stride(shape)
    else:
        if not isinstance(stride_value, list) or len(stride_value) != len(shape):
            raise ValidationError(f"{label}.stride must have the same rank as shape")
        stride = tuple(_positive_int(item, f"{label}.stride") for item in stride_value)
    required_bytes = _required_bytes(shape, stride, DTYPE_SIZES[dtype_name])
    byte_size = _unsigned(spec.get("byte_size", required_bytes), f"{label}.byte_size")
    if byte_size < required_bytes:
        raise ValidationError(f"{label}.byte_size is smaller than the declared shape and stride")
    return (
        name_offset,
        DTYPE_IDS[dtype_name],
        LOCATION_IDS[location_name],
        IO_IDS[io_kind],
        len(shape),
        0,
        byte_size,
        shape,
        stride,
    )


def _compile_argument(
    argument: Any,
    names: dict[str, int],
    device_names: set[str],
    arena_bytes: int,
    weight_size: int,
    label: str,
) -> tuple[int, int, int, int]:
    if not isinstance(argument, dict) or len(argument) != 1:
        raise ValidationError(f"{label} must contain exactly one argument kind")
    kind, raw_value = next(iter(argument.items()))
    if kind not in ARG_IDS:
        raise ValidationError(f"{label} has unsupported kind {kind!r}")
    if kind == "tensor":
        if raw_value not in names:
            raise ValidationError(f"{label} references unknown tensor {raw_value!r}")
        if raw_value not in device_names:
            raise ValidationError(f"{label} references a host tensor; Kernel arguments must be device tensors")
        return ARG_IDS[kind], names[raw_value], 0, 0
    if kind == "scalar_u32":
        return ARG_IDS[kind], 0, 0, _unsigned(raw_value, label, maximum=2**32 - 1)
    if kind == "scalar_f32":
        if isinstance(raw_value, bool) or not isinstance(raw_value, (int, float)) or not math.isfinite(raw_value):
            raise ValidationError(f"{label} must be a finite float")
        try:
            bits = struct.unpack("<I", struct.pack("<f", float(raw_value)))[0]
        except OverflowError as exc:
            raise ValidationError(f"{label} is outside the FP32 range") from exc
        return ARG_IDS[kind], 0, 0, bits
    offset = _unsigned(raw_value, label)
    limit = arena_bytes if kind == "arena_offset" else weight_size
    if offset >= limit:
        raise ValidationError(f"{label} offset {offset} is outside its {limit}-byte allocation")
    return ARG_IDS[kind], 0, offset, 0


def _validate_graph_templates(value: Any, dispatches: list[Any]) -> set[str]:
    if not isinstance(value, list):
        raise ValidationError("cuda_graph_templates must be a list")
    known = {
        item.get("profile")
        for item in dispatches
        if isinstance(item, dict) and isinstance(item.get("profile"), str)
    }
    result: set[str] = set()
    for index, item in enumerate(value):
        if not isinstance(item, dict) or item.get("profile") not in known:
            raise ValidationError(f"cuda_graph_templates[{index}] references an unknown profile")
        profile = item["profile"]
        if profile in result:
            raise ValidationError(f"duplicate CUDA Graph template for profile {profile}")
        result.add(profile)
    return result


def _dimensions(value: Any, label: str) -> tuple[int, int, int]:
    if not isinstance(value, list) or len(value) != 3:
        raise ValidationError(f"{label} must contain exactly three dimensions")
    return tuple(_positive_int(item, label, maximum=2**32 - 1) for item in value)  # type: ignore[return-value]


def _unsigned(value: Any, label: str, maximum: int = 2**64 - 1) -> int:
    if isinstance(value, bool) or not isinstance(value, int):
        raise ValidationError(f"{label} must be an unsigned integer")
    converted = value
    if converted < 0 or converted > maximum:
        raise ValidationError(f"{label} must be an unsigned integer no greater than {maximum}")
    return converted


def _positive_int(value: Any, label: str, maximum: int = 2**63 - 1) -> int:
    result = _unsigned(value, label, maximum)
    if result == 0:
        raise ValidationError(f"{label} must be positive")
    return result


def _contiguous_stride(shape: tuple[int, ...]) -> tuple[int, ...]:
    result: list[int] = []
    value = 1
    for dimension in reversed(shape):
        result.append(value)
        value *= dimension
    return tuple(reversed(result))


def _required_bytes(shape: tuple[int, ...], stride: tuple[int, ...], item_size: int) -> int:
    last_element = sum((dimension - 1) * step for dimension, step in zip(shape, stride))
    return (last_element + 1) * item_size
